bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java

```diff
@@ -0,0 +1,329 @@
+    } catch (IOException e) {
+      updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+    } catch (ResponseException e) {
+      updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+    } catch (ConfigurationException e) {
+      updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+    } catch (AuthenticationException e) {
+      handleAuthenticationException(e);
+    }
+  }
+
+  @Override
+  public void initialize() {
+    this.setupJob = scheduler.schedule(() -> {
+      try {
+        setupDevice();
+        setupRefreshInterval();
+      } catch (ResponseException | IOException e) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+      } catch (AuthenticationException e) {
+        handleAuthenticationException(e);
+      } catch (ConfigurationException e) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+      }
+    }, 0, TimeUnit.SECONDS);
+  }
+
+  protected PJLinkDeviceConfiguration getConfiguration() throws ConfigurationException {
+    PJLinkDeviceConfiguration config = this.config;
+    try {
+      validateConfigurationParameters(getThing().getConfiguration().getProperties());
+    } catch (ConfigValidationException e) {
+      String message = e.getValidationMessages().entrySet().stream()
+          .map((Map.Entry<String, String> a) -> (a.getKey() + ": " + a.getValue())).collect(Collectors.joining("; "));
+      throw new ConfigurationException(message);
+    }
+    if (config == null) {
+      this.config = config = getConfigAs(PJLinkDeviceConfiguration.class);
+    }
+    return config;
+  }
+
+  private void clearSetupJob() {
+    ScheduledFuture<?> setupJob = this.setupJob;
+    if (setupJob != null) {
+      setupJob.cancel(true);
+      this.setupJob = null;
+    }
+  }
+
+  private void clearRefreshInterval() {
+    ScheduledFuture<?> refreshJob = this.refreshJob;
+    if (refreshJob != null) {
+      refreshJob.cancel(true);
+      this.refreshJob = null;
+    }
+  }
+
+  private void handleAuthenticationException(AuthenticationException e) {
+    updateProperty(PJLinkDeviceBindingConstants.PROPERTY_AUTHENTICATION_REQUIRED, Boolean.TRUE.toString());
+    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+  }
+
+  private void setupDevice() throws ConfigurationException, IOException, AuthenticationException, ResponseException {
+    PJLinkDevice device = getDevice();
+    device.checkAvailability();
+
+    updateDeviceProperties(device);
+    updateInputChannelStates(device);
+    updateStatus(ThingStatus.ONLINE);
+  }
+
+  private void setupRefreshInterval() throws ConfigurationException {
+    clearRefreshInterval();
+    PJLinkDeviceConfiguration config = PJLinkDeviceHandler.this.getConfiguration();
```

Comment:
You could move this tto initialize as first call (before scheduling the connect job).

Reply:
What's the advantage of doing so?

Reply:
You don't need the config as parameter in the refresh job. if the config changes the handler is disposed and re-initialized anyway, so just accessing fields is faster than parsing/casting config each time you ned a value from that.

Reply:
I think I understand your point. 

Unfortunately I have to validate the configuration and instantiate the PJLinkDeviceConfiguration at some point. I could do that as early as possible as you propose and rely on having set the field earlier.

However, if I do want to keep nullability warnings to zero, I have the problem that the compiler cannot know whether the field has been set before, even if that seems obvious for me as a developer.

So as an alternative, I have extended getConfiguration() to first check if the validation/instantiation has been done before and reuse the result:
```
protected PJLinkDeviceConfiguration getConfiguration() throws ConfigurationException {
    PJLinkDeviceConfiguration config = this.config;
    if(config != null) {
      return config;
    }
    [... instantiation and validation logic, setting this.config ...]
}
```
This should fix the very valid 'getting and parsing the config over and over again' concern you raised.

However, I still want to pass it to refresh() as a parameter, as I do not want to duplicate the exception handling in case the configuration is invalid. Even if the exception can never be raised at this point (because it was already instantiated earlier during setupDevice()), the method signature of getConfiguration() still contains it and thus it needs to be handled somehow.

Duplicating the exception handling inside of the job can be avoided by calling getConfiguration() already before of the job, so that the exception handling already in place in the stack can be reused. But then I need to pass it to the job as a parameter.